python/lsst/daf/butler/core/permissions.py

```diff
@@ -0,0 +1,124 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Code relating to control of permissions based on DatasetType."""
+
+__all__ = ("Permissions", "PermissionsValidationError", "PermissionsConfig")
+
+import logging
+from .config import Config
+from .configSupport import LookupKey, processLookupConfigList
+from .exceptions import ValidationError
+
+log = logging.getLogger(__name__)
+
+
+class PermissionsValidationError(ValidationError):
+    """Exception thrown when a permissions list is not consistent with the
+    associated `DatasetType`."""
+    pass
+
+
+class PermissionsConfig(Config):
+    """Configuration information for `Permissions`"""
+    pass
+
+
+class Permissions:
+    """Control whether a `DatasetType` is allowed to be handled.
+
+    Parameters
+    ----------
+    config : `PermisssionsConfig` or `str`
+        Load configuration.  If `None` then this is equivalent to having
+        no restrictions.
+    """
+
+    matchAllKey = LookupKey("all")
+    """Configuration key associated with matching everything."""
+
+    def __init__(self, config):
+        # Default is to accept all and reject nothing
+        self._accept = set()
+        self._reject = set()
+
+        if config is not None:
+            self.config = PermissionsConfig(config)
+
+            if "accept" in self.config:
+                self._accept = processLookupConfigList(self.config["accept"])
+            if "reject" in self.config:
+                self._reject = processLookupConfigList(self.config["reject"])
+
+        if self.matchAllKey in self._accept and self.matchAllKey in self._reject:
+            raise PermissionsValidationError("Can not explicitly accept 'all' and reject 'all'"
+                                             " in one configuration")
+
+    def hasPermission(self, entity):
+        """Check whether the supplied entity has permission for whatever
+        this `Permissions` class is associated with.
+
+        Parameters
+        ----------
+        entity : `DatasetType`, `DatasetRef`, or `StorageClass`
+            Instance to use to look in permissions table.
+            The entity itself reports the `LookupKey` that is relevant.
+
+        Returns
+        -------
+        allowed : `bool`
+            `True` if the entity is allowed.
+        """
+
+        # Get the names to use for lookup
+        names = set(entity._lookupNames())
+
+        # Test if this entity is explicitly mentioned for accept/reject
+        isExplicitlyAccepted = bool(names & self._accept)
+
+        if isExplicitlyAccepted:
+            return True
+
+        isExplicitlyRejected = bool(names & self._reject)
+
+        if isExplicitlyRejected:
+            return False
+
+        # Now look for wildcard match -- we have to also check for dataId
+        # overrides
```

Comment:
Does the earlier short-circuiting mean that you can't accept or reject at the DatasetType level and then reverse that at the dimension level?  I don't actually know that we need that, but I think it's how I would have expected things to work (depending on the form of the config, which I haven't internalized, and we might want to document that if it *doesn't* work.

Reply:
I will take another look at the dimensions side of things. If you accept a specific datasettype but have explicit reject on the dimensions then you will accept it. I think that if you accept dimensions but reject a datasettype then the datasettype should be rejected but I will see if I have a test for that.

Reply:
As discussed on slack, we can already encode "accept only `raw` from `LATISS`" so I'm resolving this conversation.